handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,463 @@
+        testClientException(SslProvider.OPENSSL_REFCNT);
+    }
+
+    /**
+     * Testing what happens if the {@link ClientOcspContext} throws an {@link Exception}.
+     *
+     * The exception should bubble up on the client side and the connection should get closed.
+     */
+    private void testClientException(SslProvider sslProvider) throws Exception {
+        final AtomicReference<Throwable> causeRef = new AtomicReference<Throwable>();
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        ChannelInboundHandlerAdapter clientHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    causeRef.set(cause);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        byte[] response = newOcspResponse();
+        OcspClientCallback callback = new OcspClientCallback() {
+            @Override
+            public boolean verify(byte[] response) throws Exception {
+                throw new OcspTestException("testClientException");
+            }
+        };
+
+        handshake(sslProvider, latch, null, response, clientHandler, callback);
+
+        Throwable cause = causeRef.get();
+        assertTrue("" + cause, isCausedBy(cause, OcspTestException.class));
+    }
+
+    private static void handshake(SslProvider sslProvider, CountDownLatch latch, ChannelHandler serverHandler,
+            byte[] response, ChannelHandler clientHandler, OcspClientCallback callback) throws Exception {
+
+        SelfSignedCertificate ssc = new SelfSignedCertificate();
+        try {
+            SslContext serverSslContext = SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey())
+                    .sslProvider(sslProvider)
+                    .build();
+            ((ReferenceCountedOpenSslContext) serverSslContext).enableOcsp();
+
+            try {
+                SslContext clientSslContext = SslContextBuilder.forClient()
+                        .sslProvider(sslProvider)
+                        .trustManager(InsecureTrustManagerFactory.INSTANCE)
+                        .build();
+                ((ReferenceCountedOpenSslContext) clientSslContext).enableOcsp();
+
+                try {
+                    EventLoopGroup group = new DefaultEventLoopGroup();
+                    try {
+                        LocalAddress address = new LocalAddress("handshake-" + Math.random());
+                        Channel server = newServer(group, address, serverSslContext, response, serverHandler);
+                        Channel client = newClient(group, address, clientSslContext, callback, clientHandler);
+                        try {
+                            assertTrue("Something went wrong.", latch.await(10L, TimeUnit.SECONDS));
+                        } finally {
+                            client.close();
```

Comment:
consider blocking on the channel close futures. If the event loop is closed before the channels close we could leak buffers (close may execute a task on the event loop to cleanup internal queues).